examples/fluids/src/petsc_ops.c

```diff
@@ -425,78 +425,83 @@ PetscErrorCode ApplyAddCeedOperatorLocalToLocal(Vec X_loc, Vec Y_loc, OperatorAp
+ * @param[in]  assemble Whether to assemble `Amat` and `Pmat` if they are not `mat_ceed`
+ * @param[out] Amat     `Mat` to be used for the solver `Amat`
+ * @param[out] Pmat     `Mat` to be used for the solver `Pmat`
+ */
+PetscErrorCode CreateSolveOperatorsFromMatCeed(KSP ksp, Mat mat_ceed, PetscBool assemble, Mat *Amat, Mat *Pmat) {
+  PetscBool use_matceed_pmat, assemble_amat = PETSC_FALSE;
+  MatType   mat_ceed_inner_type;
 
   PetscFunctionBeginUser;
-  PetscCall(MatShellGetContext(A, &ctx));
-  PetscCall(ApplyCeedOperatorGlobalToGlobal(X, Y, ctx));
-  PetscFunctionReturn(PETSC_SUCCESS);
-};
-
-// -----------------------------------------------------------------------------
-// Returns the computed diagonal of the operator
-// -----------------------------------------------------------------------------
-PetscErrorCode MatGetDiag_Ceed(Mat A, Vec D) {
-  OperatorApplyContext ctx;
-  Vec                  Y_loc;
-  PetscMemType         mem_type;
+  PetscCall(MatCeedGetInnerMatType(mat_ceed, &mat_ceed_inner_type));
+  {  // Determine if Amat should be MATCEED or assembled
+    const char *ksp_prefix = NULL;
+
+    PetscCall(KSPGetOptionsPrefix(ksp, &ksp_prefix));
+    PetscOptionsBegin(PetscObjectComm((PetscObject)mat_ceed), ksp_prefix, "", NULL);
+    PetscCall(PetscOptionsBool("-matceed_assemble_amat", "Assemble the A matrix for KSP solve", NULL, assemble_amat, &assemble_amat, NULL));
+    PetscOptionsEnd();
+  }
 
-  PetscFunctionBeginUser;
-  PetscCall(MatShellGetContext(A, &ctx));
-  Ceed ceed = ctx->ceed;
-  if (ctx->Y_loc) Y_loc = ctx->Y_loc;
-  else PetscCall(DMGetLocalVector(ctx->dm_y, &Y_loc));
+  if (assemble_amat) {
+    PetscCall(MatConvert(mat_ceed, mat_ceed_inner_type, MAT_INITIAL_MATRIX, Amat));
+    if (assemble) PetscCall(MatCeedAssembleCOO(mat_ceed, *Amat));
 
-  // -- Place PETSc vector in libCEED vector
-  PetscCall(VecP2C(Y_loc, &mem_type, ctx->y_ceed));
+    PetscCall(PetscObjectReference((PetscObject)*Amat));
+    *Pmat = *Amat;
+    PetscFunctionReturn(PETSC_SUCCESS);
+  } else {
+    PetscCall(PetscObjectReference((PetscObject)mat_ceed));
+    *Amat = mat_ceed;
+  }
 
-  // -- Compute Diagonal
-  PetscCall(PetscLogEventBegin(FLUIDS_CeedOperatorAssembleDiagonal, A, D, 0, 0));
-  PetscCall(PetscLogGpuTimeBegin());
-  PetscCallCeed(ceed, CeedOperatorLinearAssembleDiagonal(ctx->op, ctx->y_ceed, CEED_REQUEST_IMMEDIATE));
-  PetscCall(PetscLogGpuTimeEnd());
-  PetscCall(PetscLogEventEnd(FLUIDS_CeedOperatorAssembleDiagonal, A, D, 0, 0));
+  {  // Determine if Pmat should be MATCEED or assembled
+    PC     pc;
+    PCType pc_type;
 
-  // -- Local-to-Global
-  PetscCall(VecC2P(ctx->y_ceed, mem_type, Y_loc));
-  PetscCall(VecZeroEntries(D));
-  PetscCall(DMLocalToGlobal(ctx->dm_y, Y_loc, ADD_VALUES, D));
+    PetscCall(KSPGetPC(ksp, &pc));
+    PetscCall(PCGetType(pc, &pc_type));
+    PetscCall(PetscStrcmpAny(pc_type, &use_matceed_pmat, PCJACOBI, PCVPBJACOBI, PCPBJACOBI, ""));
```

Comment:
Are there other common PCs that you use here that we could try to get MatCEED to support? I think AMG would be hard to support, but other MATOPS are easier to support

Reply:
I honestly not sure what "common" PCs are used. VPBJacobi handles pretty much all of our use cases. Otherwise ILU is nice, but I think we've already established supporting that would require changes to how that is implemented in PETSc.

Other random musings from looking around haphazardly:

- [`PCKACZMARZ`](https://petsc.org/release/manualpages/PC/PCKACZMARZ/) requires `MatGetRow`. Shouldn't be too difficult to implement, but we'd need to be careful about when to assemble (currently the assembly is agressive, on every call to `MatGetDiagonal` and friends)
- [`PCASM`](https://petsc.org/release/manualpages/PC/PCASM/) Requires `MatCreateSubMatrices` (not sure if we have that), `MatIncreaseOverlap`

Reply:
PC Jacobi knows to keep the diagonal until the mat changes, so we're fine there with costs.

MatCreateSubMatrices wouldn't be easy to enable, but `MatGetRow` wouldn't be hard. I'd assemble once into the inner full AIJ and then only reassemble if the mat state changes between calls to `MatGetRow`

Reply:
Yeah, I think we'd need internally keep track of the the `MATCEED`'s object state. Then whenever `MatGet*` is called, [check if the state has changed ](https://petsc.org/release/manualpages/Sys/PetscObjectStateGet/#petscobjectstateget) since the last internal assembly. Then we'd indicate that internal reassembly needs to be redone via `MatAssemblyBegin/End`.

Reply:
Yeah, I had similar logic floating around somewhere but I'm not sure where it went or why